lib/phoenix_live_view.ex

```diff
@@ -740,9 +740,45 @@ defmodule Phoenix.LiveView do
   also uses `assign_new/3` to fetch the `:current_user` in its `mount/3` callback,
   as above, the assign will be fetched from the parent LiveView, once again
   avoiding additional database queries.
+
+  Note that `fun` also provides access to the previously assigned values:
+
+      assigns =
+          assigns
+          |> assign_new(:foo, fn -> "foo" end)
+          |> assign_new(:bar, fn %{foo: foo} -> foo <> "bar" end)
   '''
   def assign_new(socket_or_assigns, key, fun)
 
+  def assign_new(%Socket{} = socket, key, fun) when is_function(fun, 1) do
+    validate_assign_key!(key)
+
+    case socket do
+      %{assigns: %{^key => _}} ->
+        socket
+
+      %{private: %{assign_new: {assigns, keys}}} ->
+        # It is important to store the keys even if they are not in assigns
+        # because maybe the controller doesn't have it but the view does.
+        socket = put_in(socket.private.assign_new, {assigns, [key | keys]})
+
+        # take only public assigns from `socket.assigns`
+        new_assigns =
+          socket.assigns
+          |> Map.take(Map.keys(socket.assigns.__changed__))
+          |> Map.merge(assigns)
+
+        Phoenix.LiveView.Utils.force_assign(
+          socket,
+          key,
+          Map.get_lazy(assigns, key, fn -> fun.(new_assigns) end)
```

Comment:
```suggestion

        Phoenix.LiveView.Utils.force_assign(
          socket,
          key,
          case assigns do
            %{^key => value} -> value
            %{} -> fun.(socket.assigns)
          end
        )
```

Reply:
We don't want to merge those assigns because the assigns in `assigns_new` are assigns that have not been moved into the socket yet.